src/classes/CRLP_ApiService_TEST.cls

```diff
@@ -0,0 +1,159 @@
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2019
+* @group API
+* @description Unit Tests for the Customizable Rollups Api Service class
+*/
+@IsTest(IsParallel=true)
+private class CRLP_ApiService_TEST {
+
+    /**
+     * @description Validate the getRollupDefsForRollupType returns the expected collection of CRLP_Rollup instances.
+     * Calls the mocking utilities to build a bunch of rollup definitions and filter groups/rules fo
+     */
+    @IsTest
+    private static void shouldReturnRollupDefinitionsWhenParameterIsValid() {
+        mockRollupDefinitions();
+
+        CRLP_ApiService service = new CRLP_ApiService();
+
+        Test.startTest();
+        for (CRLP_RollupProcessingOptions.RollupType rollupType : CRLP_RollupProcessingOptions.RollupType.values()) {
+            List<CRLP_Rollup> rollups = service.getRollupDefinitions(rollupType.name());
+            System.assertEquals(1, rollups.size(), 'The collection of returned rollups should contain exactly 1 instance for ' + rollupType.name());
+            System.assertEquals(2, rollups[0].rollupsMdt.size(), 'There should be two Rollup__mdt records associated with the CRLP_Rollup instance');
+        }
+    }
+
+    /**
+     * @description Validate the getRollupDefsForRollupType returns an exception if the RollupType is null or invalid
+     */
+    @IsTest
+    private static void shouldThrowExceptionWhenParameterIsInvalid() {
+        CRLP_ApiService crlpApiSvc = new CRLP_ApiService();
+
+        Test.startTest();
+        for (String param : new String[] { null, 'NonExistentType' }) {
+            Exception actualException = null;
+            try {
+                crlpApiSvc.getRollupDefinitions(param);
+            } catch (Exception e) {
+                actualException = e;
+            }
+            System.assertNotEquals(null, actualException, 'An exception should be thrown for invalid param: ' + param);
+            System.assert(actualException.getMessage().containsIgnoreCase(CRLP_ApiService.PARAM_ROLLUPTYPE),
+                    'Message should contain "' + CRLP_ApiService.PARAM_ROLLUPTYPE + '" param: ' + actualException.getMessage());
+        }
+    }
+
+    /**
+     * @description Mock the Rollup__mdt records for use within this unit test class
+     */
+    static void mockRollupDefinitions() {
+
+        String filterGroupId1 = CMT_UnitTestData_TEST.getNewRecordId();
```

Comment:
- Since we do not have another `filterGroupId` can we remove `1` from the name?
- Does CRLP have a test API  that we can use to extract the `rollupJSON` instead of creating it every time it is required to test rollups? Also, if someone new to rollup functionality and code would need to write a test method to create data, is there a guide on which fields can be used for which rollup definition? Having a CRLP test API would be greatly of benefit.

Reply:
There isn't an API to create these. The best I have is the `CMT_UnitTestData_TEST.createRollupRecord` method, which as you point out is complex and has a lot of parameters. It would be a lot easier if we could just instantiate `Rollup__mdt` records, but that's not yet supported when there are Entity Relationship fields on the CMT object.

What would be a lot more readable is to rebuild the `CMT_UnitTestData_TEST.createRollupRecord` logic as more of a class structure:

```
List<Rollup__mdt> = new List<Rollup__mdt>{
   CMT_UnitTestData_TEST.newRollup('Total Donations')
     .withSummaryField('Contact.npo02__TotalOppAmount__c')
     .withFilterGroupId(filterGroupId)
     .withDetailField('Opportunity.Amount')
     .withAmountField('Opportunity.Amount')
     .withDateField('Opportunity.CloseDate')
     .withOperation(CRLP_Operation.RollupType.Sum)
     .withTimeFrame(CRLP_Operation.TimeBoundOperationType.All_Time)
     .getMdtRecord()
};
```

I wouldn't mind taking that on, but I'm not sure I've got the bandwidth to redo that (and all the classes that use it) right now

Reply:
Thanks for the feedback. Perhaps the best action is to create a new WI to check into building a CRLP (or any specific domain) test API that we can call from our other test classes and keep the high level code in unit tests.